From c1816d9f48376ab76dee5642122806d106793416 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Wed, 18 Dec 2024 12:28:34 +0100
Subject: [PATCH] media: stm32: dcmipp: correct PIPEDIFF handling for direct
 aux

Currently PIPEDIFF handling is done within the isp subdev.  This
cannot fully work since, when only configuring the aux pipeline
in such mode, isp subdev is not started hence PIPEDIFF configuration
point is not reached.
Put the PIPEDIFF control within the dcmipp_input subdev, related
to the enabled connection starting from the aux pipe dedicated
source pad.

Change-Id: I1999a4b659be44658563d89a10f96044cc86897c
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/425495
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
Reviewed-by: Hugues FRUCHET <hugues.fruchet@st.com>
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Alain VOLMAT <alain.volmat@st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
---
 .../st/stm32/stm32-dcmipp/dcmipp-input.c      | 12 +++++++++
 .../st/stm32/stm32-dcmipp/dcmipp-isp.c        | 27 -------------------
 2 files changed, 12 insertions(+), 27 deletions(-)

diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-input.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-input.c
index 124f4b179f25..40f432f44510 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-input.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-input.c
@@ -45,6 +45,7 @@
 #define DCMIPP_P0FSCR	0x404
 #define DCMIPP_P1FSCR	0x804
 #define DCMIPP_P2FSCR	0xC04
+#define DCMIPP_P1FSCR_PIPEDIFF		BIT(18)
 #define DCMIPP_PxFSCR_DTMODE_MASK	GENMASK(17, 16)
 #define DCMIPP_PxFSCR_DTMODE_SHIFT	16
 #define DCMIPP_PxFSCR_DTMODE_DTIDA	0x00
@@ -509,6 +510,17 @@ static int dcmipp_inp_s_stream(struct v4l2_subdev *sd, int enable)
 		if (ret)
 			goto error_s_stream;
 
+		/*
+		 * Check if the Aux pipe source pad is connected / enabled
+		 * or not.  If enabled, it means that Aux pipe works alone
+		 * and not connected to Main pipe ISP
+		 */
+		if (inp->ved.ent->num_pads >= 3 &&
+		    !media_pad_remote_pad_first(&inp->ved.pads[3]))
+			reg_clear(inp, DCMIPP_P1FSCR, DCMIPP_P1FSCR_PIPEDIFF);
+		else
+			reg_set(inp, DCMIPP_P1FSCR, DCMIPP_P1FSCR_PIPEDIFF);
+
 		ret = v4l2_subdev_call(s_subdev, video, s_stream, enable);
 		if (ret < 0) {
 			dev_err(inp->dev,
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
index 60aacda22546..a4c70e4b5365 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
@@ -29,9 +29,6 @@
 
 #define DCMIPP_CMSR2_P1VSYNCF BIT(18)
 
-#define DCMIPP_P1FSCR	0x804
-#define DCMIPP_P1FSCR_PIPEDIFF BIT(18)
-
 #define DCMIPP_P1SRCR	0x820
 #define DCMIPP_P1SRCR_LASTLINE_SHIFT	0
 #define DCMIPP_P1SRCR_LASTLINE_MASK	GENMASK(11, 0)
@@ -518,24 +515,6 @@ static void dcmipp_isp_config_demosaicing(struct dcmipp_isp_device *isp,
 	reg_set(isp, DCMIPP_P1DMCR, val);
 }
 
-static bool dcmipp_isp_is_aux_output_enabled(struct dcmipp_isp_device *isp)
-{
-	struct media_link *link;
-
-	for_each_media_entity_data_link(isp->ved.ent, link) {
-		if (link->source != &isp->ved.pads[1])
-			continue;
-
-		if (!(link->flags & MEDIA_LNK_FL_ENABLED))
-			continue;
-
-		if (!strcmp(link->sink->entity->name, "dcmipp_aux_postproc"))
-			return true;
-	}
-
-	return false;
-}
-
 static void dcmipp_isp_config_decimation(struct dcmipp_isp_device *isp,
 					 struct v4l2_rect *crop,
 					 struct v4l2_rect *compose)
@@ -581,12 +560,6 @@ static int dcmipp_isp_s_stream(struct v4l2_subdev *sd, int enable)
 		compose = v4l2_subdev_state_get_compose(state, 0);
 		v4l2_subdev_unlock_state(state);
 
-		/* Check if link between ISP & Pipe2 postproc is enabled */
-		if (dcmipp_isp_is_aux_output_enabled(isp))
-			reg_clear(isp, DCMIPP_P1FSCR, DCMIPP_P1FSCR_PIPEDIFF);
-		else
-			reg_set(isp, DCMIPP_P1FSCR, DCMIPP_P1FSCR_PIPEDIFF);
-
 		/* Configure Statistic Removal */
 		reg_write(isp, DCMIPP_P1SRCR,
 			  ((crop->top << DCMIPP_P1SRCR_FIRSTLINEDEL_SHIFT) |
-- 
2.50.1

